Improve compatibility of holochain types with Zome WASM, unify Timestamp type

pjkundert opened the pull request:
There are two implementations of the Timestamp; this merges them into one, and makes the Timestamp type available in WASM code.   The now() method was moved from being an impl for Timestamp, eg. Timestamp::now(), to be a free-floating function in the crate, eg. timestamp::now(), so code changes are minimal and easy to understand.  This improves the log output considerably, as Timestamps are always Debug-formatted as times, not as a pair of big numbers.  Also, updated the fixt plumbing to actually create more-or-less "correct" random Timestamps for testing, so that conversion to Duration and SystemTime won't fail during tests that assume sane Timestamps for comparison.

In addition, various holochain error types lacked some derive implementation, making it difficult to write Zome test code in rust (as Err types couldn't be compared).  

Finally, the holo_hash types and related functions were moved around a bit, to allow Zome code to build "wrappers" around all DHT times, allowing them to be De/Serialized as Strings for use over Zome API calls.  This code is not included here, but used heavily in hash-heavy Zomes, such as the new holofuel implementation I'm working on...

o Move Timestamp from holochain/types to zome_types
o Give more ...Error types Debug, PartialEq, ... for Rust testing
o Move some holo_hash types/impls around so they work in Zome code
o Convert Timestamp::now() impl method to timestamp::now() function In
  order for Timestamp to be WASM compatible and usable in Zome code,
  we've moved it to holochain_wasm_types, and made the holochain_types
  timestamp now() a function, instead of a Timestamp::now() method.
  Minimal code change, vastly improved Zome coding experience.
o Update base64 crate to avoid multiple inclusions

neonphog commented on crates/holochain_types/src/prelude.rs:
```diff
@@ -29,6 +29,7 @@ pub use crate::header::*;
 pub use crate::link::*;
 pub use crate::metadata::*;
 pub use crate::signal::*;
+pub use crate::timestamp; // for timestmap::now()
```
```suggestion
pub use crate::timestamp; // for timestamp::now()
```

neonphog commented on crates/holochain_zome_types/Cargo.toml:
```diff
@@ -15,6 +15,7 @@ edition = "2018"
 holo_hash = { path = "../holo_hash" }
 holochain_serialized_bytes = "=0.0.47"
 paste = "=1.0.3"
+chrono = "0.4.6"
```
@thedavidmeister - I like this PR in general - this line will pull chrono into WASM, think that is acceptable?

pjkundert replied:
Yes, that is probably the biggest question.  Presently all Zome logging includes a lot of Timestamps as eg. [1234566345,134421343], making the logging not very useful -- this would buy us at least logging that is more usable.  So, that alone might be a worthwhile investment of the compiled Zome space required.

I think most non-trivial Zomes will deal in times, so it'll usually be a wash.

pjkundert replied:
Furthermore, this dramatically cleans up the usage of the chrono DateTime in dealing with Timestamps and Durations.  Presently, they are *very* ... broken.  It is super easy to construct Timestamps that just panic in the DateTime module (fixt did it a lot).  That is all solved with this implementation (I have exhaustive tests in my mutualcredit Zome, when we can move into holochain if we wish.)

pjkundert replied:
Just working on rebasing this forward to current develop...

thedavidmeister replied:
yes i'm happy with the size changes, seems the compiler is good about eliding it if you're not using it, and if you're using it then you're using it :)

pjkundert commented:
OK, rebased, fixed suggestion.

I really think this is a worthwhile and valuable patch; really improves the Timestamp reliability, usability and consistency.

Thanks for considering it!

thedavidmeister commented:
@pjkundert @neonphog just a first question is that this introduces a dependency, can we get a smoke test on the overhead it adds to wasm?

compile a wasm file with this in it and without this in it, in release profile, and use `wasm-opt` to strip out excess

```
# install wasm-opt
apt-get install binaryen
# optimise the wasm
wasm-opt --strip-debug -Oz input_file.wasm -o output_file.wasm
```

pjkundert commented:
@thedavidmeister @neonphog 

So, apparently the Rust compiler is pretty good at eliding unnecessary/unused crates from compiled WASM:

```
[perry@bullseye holochain (feature-timestamp-without %=)]$ ls -l *wasm*
-rw-rw-r-- 1 perry perry 762298 Feb  5 14:39 test_wasm_create_entry.wasm-with-chrono
-rw-rw-r-- 1 perry perry 762711 Feb  5 15:00 test_wasm_create_entry.wasm-without-chrono
```

These were compiled with and without the chrono crate included (using the `develop` branch, and my `feature-timestamp` branch), then optimized.

neonphog commented:
sooo... it's bigger *without* chrono : ) - did the wasm you built make use of some of the chrono apis?

pjkundert commented:
I suspect the size change was probably just due to one of the (few) other changed crates; the fact that the Zome didn't use any of the chrono APIs made the size pretty much the same.  I guess the logging doesn't trigger any inclusion of the APIs if the Zome doesn't use any debug! macros.

neonphog reviewed (approved):
I guess my vote is to go ahead and merge it - If we find out later that extensive use of chrono adds a large amount to WASM size, we can figure out how to make it an optional feature.

thedavidmeister commented on crates/holochain_zome_types/src/timestamp.rs:
```diff
@@ -1,16 +1,398 @@
+// from_timestamp due to out-of-range secs or nsecs, making all code using/displaying a Timestamp
+// this way dangerously fragile!
+/*
+impl From<Timestamp> for chrono::DateTime<chrono::Utc> {
```
if the intent here is to scare people away from implementing this without realising, i doubt they will ever read this comment

what about something like this?

```rust
impl From<Timestamp> for chrono::DateTime<chrono::Utc> {
  panic!("Converting from a timestamp to a DateTime without error handling is always a bug. Use TryFrom instead.");
}
```

pjkundert replied:
I think its best just to not provide the infallible conversions; I adjusted the comment to be an implementation note to someone who happens to look at the code to see why there aren't any infallible conversion.

thedavidmeister commented on crates/holochain_zome_types/src/timestamp.rs:
```diff
@@ -1,16 +1,398 @@
+    ops::{Add, Sub},
+    str::FromStr,
+};
+//use chrono::Duration;
```
remove this

thedavidmeister commented on crates/holochain_zome_types/src/timestamp.rs:
```diff
@@ -1,16 +1,398 @@
+}
+
+/*
+/// Timestamp +/- Into<chrono::Duration>: chrono::Duration allows +/- offsets.  Conflicts with
```
delete this commented code?

thedavidmeister commented on crates/holochain_zome_types/src/timestamp.rs:
```diff
@@ -1,16 +1,398 @@
+	let t2 = Timestamp( 0, 0 ) + core::time::Duration::new(0,1);//nanoseconds( 1 );//chrono::Duration::nanoseconds( 1 );
+	assert_eq!( t2,  Ok(Timestamp( 0, 1 )));
+
+	//let t3 = Timestamp( 0, 0 ) - chrono::Duration::nanoseconds( 1 );
```
delete commented code?

pjkundert commented:
Thanks, @thedavidmeister!  Done.

guillemcordoba commented:
I can confirm @pjkundert 's concerns, recently had to do some logic involving timestamps, and it was indeed quite clunky. Happy to see this introduced.

neonphog commented:
Erm... apparently CI doesn't run on forks? - running here: https://app.circleci.com/pipelines/github/holochain/holochain/4776/workflows/23338462-f8ed-4f0d-974f-171ea0b0d804/jobs/6080

neonphog commented:
CI + fmt + clippy - https://github.com/holochain/holochain/pull/635

neonphog commented:
[ci passed](https://circleci.com/gh/holochain/holochain/6081?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) - merging
